kratos/mpi/utilities/gather_modelpart_utility.cpp

```diff
@@ -0,0 +1,208 @@
+#include "mpi/utilities/gather_modelpart_utility.h"
+#include "mpi/utilities/parallel_fill_communicator.h"
+#include "includes/data_communicator.h"
+#include "mpi/includes/mpi_communicator.h"
+
+namespace Kratos
+{
+
+GatherModelPartUtility::GatherModelPartUtility(int gather_rank,
+                                               ModelPart& origin_model_part,
+                                               int mesh_id,
+                                               ModelPart& destination_model_part)
+    : mr_model_part(destination_model_part), mgather_rank(gather_rank)
+{
+    KRATOS_TRY;
+
+    const int mpi_rank = DataCommunicator::GetDefault().Rank();
+    const int mpi_size = DataCommunicator::GetDefault().Size();
+
+    destination_model_part.GetNodalSolutionStepVariablesList() =
+        origin_model_part.GetNodalSolutionStepVariablesList();
+    if (DataCommunicator::GetDefault().IsDistributed())
+    {
+      VariablesList* pVariablesList =
+          &destination_model_part.GetNodalSolutionStepVariablesList();
+      destination_model_part.SetCommunicator(
+          Communicator::Pointer(new MPICommunicator(pVariablesList)));
```

Comment:
Use the origin DataCommunicator also for the destination_model_part:
`Communicator::Pointer(new MPICommunicator(pVariablesList, r_comm)));`